File: LeetCode/Medium/0841-keys-and-rooms/0841-keys-and-rooms.java
class Solution {
    int count = 0;
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        boolean[] visited = new boolean[rooms.size()];
        dfs(0, rooms, visited);
        
        return rooms.size() == count;
    }
    
    private void dfs(int room, List<List<Integer>> rooms, boolean[] visited){
        visited[room] = true;
        ++count;

        for (Integer key : rooms.get(room)) if (!visited[key]) dfs(key, rooms, visited);
    }
}

File: LeetCode/Medium/0841-keys-and-rooms/0841-keys-and-rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        vector<bool> visited(rooms.size(), false);
        
        dfs(0, rooms, visited);

        for (int i = 1; i < rooms.size(); ++i) if (!visited[i]) return false;
        
        return true;
    }

    void dfs(int room, vector<vector<int>>& rooms, vector<bool>& visited){
        visited[room] = true;

        for (int key: rooms[room]) if (!visited[key]) {dfs(key, rooms, visited);}
    }
};